Sources/Parsers/StoryboardParser.swift

```diff
@@ -7,90 +7,214 @@
+public enum StoryboardParserError: Error, CustomStringConvertible {
+  case popoverPresentationSegueWithoutAnchorView
+  case unwindSegueWithoutUnwindAction
+  case relationshipSegueWithoutRelationship
+
+  public var description: String {
+    switch self {
+    case .popoverPresentationSegueWithoutAnchorView:
+      return "Invalid storyboard file: A popoverPresentation type segue without "
+    case .unwindSegueWithoutUnwindAction:
+      return "Invalid strings file"
+    case .relationshipSegueWithoutRelationship:
+      return "Invalid strings file"
+    }
   }
+}
 
-  struct Scene {
-    let storyboardID: String
-    let tag: String
-    let customClass: String?
-    let customModule: String?
+// swiftlint:disable type_body_length
+public final class StoryboardParser {
+  var storyboardFiles = [String: Set<Scene>]()
+  var segueDestinationIds = Set<String>()
+  var modules = Set<String>()
+
+  public struct Scene: Element {
+    let identifier: String
+    let storyboardId: String
+    let type: StoryboardParser.SceneElement
+    let isInitial: Bool
+    let reuseIdentifiers: Set<StoryboardParser.Cell>
+    let segues: Set<StoryboardParser.Segue>
   }
 
-  struct Segue {
+  public struct Segue: Element {
     let identifier: String
-    let customClass: String?
-    let customModule: String?
+    let storyboardId: String
+    let type: StoryboardParser.SegueElement
+    let destinationId: String
+    let kind: SegueKind
+    let popoverPresentationAnchorView: String?
+    let unwindAction: String?
+    let relationship: String?
   }
 
-  var initialScenes = [String: InitialScene]()
-  var storyboardsScenes = [String: Set<Scene>]()
-  var storyboardsSegues = [String: Set<Segue>]()
-  var modules = Set<String>()
+  public struct Cell: Element {
+    let identifier: String
+    let storyboardId: String
+    let type: StoryboardParser.CellElement
+    var cellType: StoryboardParser.CellType {
+      return StoryboardParser.CellType(rawValue: type.tag)!
+    }
+    var dictionary: [String:Any] {
+      return [
+        "storyboardId": storyboardId,
+        "tag": type.tag,
+        "class": type.klass,
+        "module": type.module,
+        "type": type.type
+      ]
+    }
+  }
 
   public init() {}
 
   private class ParserDelegate: NSObject, XMLParserDelegate {
-    var initialViewControllerObjectID: String?
-    var initialScene: InitialScene?
+    private var initialViewControllerObjectID: String?
+    private var storyboardPlatform: StoryboardParser.Platform = .iOS
     var scenes = Set<Scene>()
-    var segues = Set<Segue>()
-    var inScene = false
-    var readyForFirstObject = false
-    var readyForConnections = false
+    var segueDestinationIds = Set<String>()
+    var modules = Set<String>()
+    private var inScene = false
+    private var readyForFirstObject = false
+    private var readyForConnections = false
+    private var readyForTableView = false
+    private var readyForCollectionView = false
```

Comment:
At this point I think it would be worth migrating the XML parsing throughout the SwiftGenKit framework to a DOM parser (like [SWXMLHash](https://github.com/drmohundro/SWXMLHash) for example) to make the logic more readable and maintainable.

In the first versions of SwiftGen I decided to use the built-in SAX Parser in iOS (`NSXMLParser`) because I didn't want to introduce an additional dependency while the parsing logic was simple enough. But with the use cases growing and the different use cases we want to handle becoming wider, it might be time to migrate now.